File: app/local_plugins/assistant_server/android/src/main/kotlin/com/aiagent/assistant_server/AssistantServerCore.kt
package com.aiagent.assistant_server

import android.content.Context
import android.util.Log
import kotlinx.coroutines.*

/**
 * AI 助理场景的总编排器（进程内单例）。
 *
 * 互斥：[active] 至多一个会话；新 start 调用时若已有 active 则抛 `assistant.session_busy`。
 */
internal object AssistantServerCore {

    private const val TAG = "AssistantServerCore"

    @Volatile private var active: AssistantSession? = null
    @Volatile private var emit: ((Map<String, Any?>) -> Unit)? = null

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    /** 由 Plugin 在 onAttached 时注入；EventChannel 转发器。 */
    fun bindEmitter(emitter: (Map<String, Any?>) -> Unit) {
        this.emit = emitter
    }

    fun unbindEmitter() {
        this.emit = null
    }

    fun activeSessionId(): String? = active?.sessionId

    /**
     * 启动 AI 助理会话。返回 sessionId。
     * 失败 → 抛 IllegalStateException + emit error/sessionState=error。
     */
    fun startAssistant(
        sessionId: String,
        request: AssistantRequest,
        context: Context,
    ): String {
        synchronized(this) {
            active?.let {
                throw IllegalStateException("assistant.session_busy: another session ${it.sessionId} is active")
            }
            val emitter: (Map<String, Any?>) -> Unit = emit ?: { evt ->
                Log.w(TAG, "no emitter bound; event dropped: $evt")
            }
            val session = AssistantSession(sessionId, context, request, emitter)
            active = session

            emitter(AssistantEvents.sessionState(sessionId, "starting"))
            scope.launch {
                try {
                    session.start()
                } catch (e: CancellationException) {
                    throw e
                } catch (e: Exception) {
                    Log.e(TAG, "session start failed: ${e.message}", e)
                    val msg = e.message ?: e::class.java.simpleName
                    val code = msg.substringBefore(":").trim().takeIf {
                        it.startsWith("assistant.")
                    } ?: "assistant.start_failed"
                    session.markError(code, msg)
                    if (active === session) active = null
                }
            }
            return sessionId
        }
    }

    /**
     * 停止当前 active session。无 active 时 no-op。
     *
     * 立即把 [active] 置 null（避免重复触发 / 占位竞态），把实际的 [AssistantSession.stop]
     * 拉到 IO scope 跑：里面会 release agent runner，串行化释放在主线程上会触发 ANR。
     */
    fun stopActive() {
        val target = synchronized(this) {
            val s = active ?: return
            active = null
            s
        }
        scope.launch { runCatching { target.stop() } }
    }
}

File: apps/client/local_plugins/assistant_server/ios/Classes/AssistantServerCore.swift
import Flutter
import Foundation

/// AI 助理场景的总编排器（iOS 进程内单例）—— 与 Android `AssistantServerCore` 对齐。
///
/// 互斥：[active] 至多一个会话；新 start 调用时若已有 active 则抛 `assistant.session_busy`。
final class AssistantServerCore {

    static let shared = AssistantServerCore()
    private init() {}

    private let lock = NSLock()
    private var active: AssistantSession?
    private var emit: (([String: Any?]) -> Void)?

    private var deviceMethod: FlutterMethodChannel?

    /// 由 Plugin 在 onAttached 时注入；EventChannel 转发器 + device_jieli MethodChannel。
    func bindContext(messenger: FlutterBinaryMessenger,
                     emitter: @escaping ([String: Any?]) -> Void) {
        lock.lock()
        self.emit = emitter
        if self.deviceMethod == nil {
            self.deviceMethod = FlutterMethodChannel(
                name: "device_jieli/method",
                binaryMessenger: messenger
            )
        }
        lock.unlock()
    }

    func unbindEmitter() {
        lock.lock(); emit = nil; lock.unlock()
    }

    func activeSessionId() -> String? {
        lock.lock(); defer { lock.unlock() }
        return active?.sessionId
    }

    /// 启动 AI 助理会话。返回 sessionId。
    func startAssistant(sessionId: String, request: AssistantRequest) throws -> String {
        lock.lock()
        if let a = active {
            lock.unlock()
            throw NSError(
                domain: "assistant_server", code: -1,
                userInfo: [NSLocalizedDescriptionKey:
                    "assistant.session_busy: another session \(a.sessionId) is active"]
            )
        }
        guard let emitter = emit, let dm = deviceMethod else {
            lock.unlock()
            throw NSError(
                domain: "assistant_server", code: -2,
                userInfo: [NSLocalizedDescriptionKey: "assistant.not_bound: emit/messenger missing"]
            )
        }
        let session = AssistantSession(
            sessionId: sessionId,
            request: request,
            deviceMethod: dm,
            emit: emitter
        )
        active = session
        lock.unlock()

        emitter(AssistantEvents.sessionState(sessionId: sessionId, state: "starting"))
        session.start()

        // 监听 stopped 时把 active 清空 —— 通过观察 sessionState 事件
        // 实际是 AssistantSession.stop() 内部会 emit sessionState=stopped，
        // 这里不另做拦截；下一次 startAssistant 时如果 active 不是 stopped 的，
        // 会按 session_busy 拒绝。简单起见，给 stop 路径加 clearIfMatches。
        return sessionId
    }

    /// 停止当前 active session。无 active 时 no-op。
    func stopActive() {
        lock.lock()
        let target = active
        active = nil
        lock.unlock()
        target?.stop()
    }
}
